### PR TITLE
Incentivized function calls through Engine

### LINKED ISSUE #892: Avoid frontrunning on Participation.executeRequest

### LINKED ISSUE #898: Incentive change while there is an open request leads to unexpected result
Example scenario:

1. Registry has `10 finney` as default incentive
2. Alice calls `requestInvestment`, sending `10 finney` to the participation contract as an incentive
3. `Registry.owner` calls `setIncentive(1 eth)`
4. After this, trying to execute Alice's request either:
 a. fails if there is not enough eth in Participation, or
 b. drains more eth than expected from Participation, and potentially causes other users' requests to fail

One fix would be to attach incentive amount to the request struct in the `requestIncentive` tx

### PR DESCRIPTION
**Pull request by travs**
This fixes #892 
Incentivized requests can now be executed either (a) via the Engine by anyone through a Fund, or (b) by the request owner.

In (a), the incentive is paid in WETH to the executing Fund, and a corresponding amount of MLN is taken from the Fund and burned.
In (b), the incentive is returned to the request owner in ETH.

This PR also fixes #898 , so incentives are now tracked per request, in case some change happens to the incentive amount in the interim.

### CODE REVIEW

**Comment by travs** [comment]
Ultimately this paradigm of using `callOnExchange` for these types of function calls should change, but rather than change that architecture here, it's implemented on `EngineAdapter`. We should be able to port it over when we write this new type of external function calls for integrating lending protocols, etc.

File: src/engine/Engine.sol [hunk 1]
```diff
@@ -150,6 +177,28 @@ contract Engine is DSMath {
         emit Burn(mlnAmount);
     }
 
+    /// @dev MLN must be approved first
+    function executeRequestAndBurnMln(
```
[the remark this answers is not in the document]
Comment (SeanJCasey) [reply src/engine/Engine.sol]: I see why it's done this way after reviewing the entire PR. I think this is fine for now. If we have / imagine having more than one incentive-payable function we want executable on the EngineAdapter, maybe we can have `executeInvestmentRequestAndBurnMln() incentivePayable`, `mySecondIncentiveAndMlnBurningFunction() incentivePayable`, etc that recycles the incentive and burning logic?

**Comment by jennazenk** [comment]
Would like to hold on before merging this. Have some questions and comments around the actual mechanism (a bit different from what we discussed last week), but probably better to discuss in person or on phone @travs. Can probably discuss Wednesday or Thursday

**Comment by travs** [comment]
deprecated

**Pull request closed without merging**